### ISSUE TITLE
Receive Alarm Notifications

### ISSUE THREAD

**Issue by smvdheijden** [OSGP/Platform#125]
### Story: Receive an Alarm Notification
```
As a Grid Operator
I want to be able to receive Alarm Notifications
So that I can take appropriate actions on meter alarms
```

#### Scenario: Handle an Alarm Notification Message
```
Given an Alarm Notification Message from a device with id <<<deviceid>>>
And the device with id <<<deviceid>>> is known
When the Alarm Notification Message is received
And Push Setup Alarm should be pushed by the meter to OSGP
And Push Setup Alarm object should contain the value of Alarm Register
And Push Setup Alarm object should contain the deviceid
And Push Setup Alarm object should contain the TCP message 
(needed to setup a connection with the meter)
And the Alarm Register should be stored
And the Alarm register should be sent to the client/ integration layer (SMA)
```

#### Scenario: handle Alarm Notification for known meter
```
Given a meter with id 54321 is known in OSGP
When an Alarm Notification is received from meter 54321
Then store the Alarm Notification in the received messages
and create an OSGP Alarm Notification created from the device's Alarm Notification information
and create an OSGP Notification containing a correlation UID to enable retrieval of the OSGP Alarm Notification
and send the OSGP Notification to a configurable notification destination
```

#### Scenario: handle Alarm Notification for unknown meter
```
Given a meter with id 54321 is unknown in OSGP
When an Alarm Notification is received from meter 54321
Then log a warning for communication from an unknown meter in the application logging
and end the communication channel to the meter
and do not process the Alarm Notification
```

#### Explanation
```
Meter sends a Wake up Notification Message (WNM) to OSGP (listener). 
This WNM is a TCP/IP message containing the 
Equipment identifier of the device. 
OSGP will extract the device’s IP address from 
the message and communicates with the Equipment.

When an Alarm Notification is sent to OSGP,  the
meter pushes a Push Setup Alarm object (0-1:25.9.0.255).
This COSEM object contains also the value of the Alarm Register. 
The received Alarm Register has to be sent to a client (the Integration Layer).
```

#### Device interface
```
See DLMS Blue book:
COSEM object: Push Setup Alarm (class_id: 40, version: 0)
Attribute: Push_object_list
```

#### Input
```
- Device ID
- Push Setup Alarm object
- IP address (TCP parameters)
```
 
#### Output
```
Just a TCP level acknowledgement (ACK) of the received message.
```

To refine (questions from @rlemmers)
- Do we have Alarm Notifications for both E meters and G meters? Or just for G meters? If both E meters and G meters have these, do we need to distinguish between those in the scenarios?

**Comment by rlemmers** [comment]
Assigned this one to @bvdzwet, because he will contact me about the notification mechanism (and perhaps about other stuff as well).

**Comment by bvdzwet** [comment]
Some information of things I looked in to is included as a comment with Platform#241.
